# Newton-Raphson Method

In [2]:
import pandas as pd
import statsmodels.api as sm

# Data as given
data = {
    'Income': [50, 30, 40, 25, 60],
    'Travel_Time': [15, 25, 20, 30, 10],
    'Travel_Cost': [5, 4, 6, 3, 7],
    'Mode_Choice': [1, 2, 1, 2, 3]  # 1=Car, 2=Bus, 3=Taxi
}

df = pd.DataFrame(data)

# Create binary outcomes for pairwise comparisons

# 1. Car vs Not-Car (1 = Car, 0 = Bus or Taxi)
df['Car_Choice'] = (df['Mode_Choice'] == 1).astype(int)

# 2. Bus vs Not-Bus (1 = Bus, 0 = Car or Taxi)
df['Bus_Choice'] = (df['Mode_Choice'] == 2).astype(int)

# 3. Taxi vs Not-Taxi (1 = Taxi, 0 = Car or Bus)
df['Taxi_Choice'] = (df['Mode_Choice'] == 3).astype(int)

# Now we will run a Probit regression for each of these binary outcomes:

# Function to fit Probit model and print the results
def fit_probit_model(y, X):
    X = sm.add_constant(X)  # Add constant (intercept)
    model = sm.Probit(y, X)
    result = model.fit()
    return result

# 1. Car vs Not-Car (binary outcome)
X_car = df[['Income', 'Travel_Time', 'Travel_Cost']]
y_car = df['Car_Choice']
result_car = fit_probit_model(y_car, X_car)
print("Car vs. Not-Car Probit Model:")
print(result_car.summary())

# 2. Bus vs Not-Bus (binary outcome)
X_bus = df[['Income', 'Travel_Time', 'Travel_Cost']]
y_bus = df['Bus_Choice']
result_bus = fit_probit_model(y_bus, X_bus)
print("\nBus vs. Not-Bus Probit Model:")
print(result_bus.summary())

# 3. Taxi vs Not-Taxi (binary outcome)
X_taxi = df[['Income', 'Travel_Time', 'Travel_Cost']]
y_taxi = df['Taxi_Choice']
result_taxi = fit_probit_model(y_taxi, X_taxi)
print("\nTaxi vs. Not-Taxi Probit Model:")
print(result_taxi.summary())


Optimization terminated successfully.
         Current function value: 0.554518
         Iterations 28
Car vs. Not-Car Probit Model:
                          Probit Regression Results                           
Dep. Variable:             Car_Choice   No. Observations:                    5
Model:                         Probit   Df Residuals:                        1
Method:                           MLE   Df Model:                            3
Date:                Thu, 02 Jan 2025   Pseudo R-squ.:                  0.1761
Time:                        09:50:43   Log-Likelihood:                -2.7726
converged:                       True   LL-Null:                       -3.3651
Covariance Type:            nonrobust   LLR p-value:                    0.7566
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
const         152.2115   7.52e+07   2.03e-06      1.000   -1.47e+08    1.47

C:\ProgramData\anaconda3\Lib\site-packages\statsmodels\discrete\discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
C:\ProgramData\anaconda3\Lib\site-packages\statsmodels\discrete\discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
C:\ProgramData\anaconda3\Lib\site-packages\statsmodels\discrete\discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
C:\ProgramData\anaconda3\Lib\site-packages\statsmodels\discrete\discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
C:\ProgramData\anaconda3

# Explanation of the Results:
## Coefficients: These represent the change in the log-odds of choosing a particular mode over the others. For example, in the Car vs. Not-Car model:

###   Income coefficient: A positive coefficient means that as income increases, the likelihood of choosing "Car" over "Not-Car" increases.
###   Travel Time: A negative coefficient suggests that more travel time reduces the likelihood of choosing "Car" over "Not-Car".
###   Travel Cost: A positive coefficient suggests that higher travel costs increase the likelihood of choosing "Car" over "Not-Car".
###   Significance: The p-values (P>|z|) tell you whether the coefficients are significantly different from zero. A p-value below 0.05 typically means the coefficient is significant. For example, in the Car vs. Not-Car model, none of the variables are statistically significant at the 5% level.

###   Confidence Intervals: The [0.025, 0.975] columns show the 95% confidence intervals for the coefficients. If the confidence interval includes zero, the coefficient is not significantly different from zero at the 5% level.